# 01. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Original Data

In [2]:
# 2a) Import order data
df_ords=pd.read_csv(r'C:\Users\ponni\Achievement 4\02 Data\Original data\orders.csv',index_col=False)
# 2b) Import product data
df_prods=pd.read_csv(r'C:\Users\ponni\Achievement 4\02 Data\Original data\products.csv',index_col=False)

# 03. Import Prepared data

In [3]:
# 3a) Import wrangled order data
df_ords_prep=pd.read_csv(r'C:\Users\ponni\Achievement 4\02 Data\Prepared data\orders_wrangled.csv',index_col=False)

# 04. Test Dataframe - Data consistency check

In [4]:
# Create a data frame
df_test=pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix']=['a','b',1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
df_test['mix'] = df_test['mix'].astype('str')

In [9]:
df_test.dtypes

mix    object
dtype: object

In [10]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Create a subset with only missing value
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# Create a new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
# Checking for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
# Create a new dataframe without duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

# 05. Task 4.5

In [20]:
# Q1 Consistency check for df_prods
# ALready done

In [21]:
# Q2 describe function on your df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

All the values in the prices column seems to be consistent except the max price. It seems to be a error.

In [22]:
# Checking for max price
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

So, we can understand from the above table that the milk price was wrongly entered. Therfore, further investigation is required.

In [36]:
# Q3 Check mixed-type data for df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed data types found.

In [37]:
# Q5 Check for missing values for df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In the column days_since_prior_order has 206209 missing values.

In [43]:
# Create a subset with only missing value for df_ords
df_nan_ords=df_ords[df_ords['days_since_prior_order'].isnull()==True]
df_nan_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [49]:
df_nan_ords.order_number.describe()

count    206209.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: order_number, dtype: float64

Here, from the table above we can infer that for all the missing values the order_number is 1. The first time customers are indicated by order_number 1. 

In [48]:
# Q6 Address the missing values using an appropriate method
df_ords.days_since_prior_order.describe()

count    3.214874e+06
mean     1.111484e+01
std      9.206737e+00
min      0.000000e+00
25%      4.000000e+00
50%      7.000000e+00
75%      1.500000e+01
max      3.000000e+01
Name: days_since_prior_order, dtype: float64

In [50]:
df_ords.days_since_prior_order.median()

7.0

In [51]:
df_ords.days_since_prior_order.mean()

11.114836226863012

In [52]:
# Imputing flag value of '999' for all Nan entries
df_ords['days_since_prior_order'].fillna(999, inplace=True)

I imputed all the missing value for the column with the flag value '999'. So that, it doesnt affect the analysis results

In [54]:
# Checking for missing values
df_ords.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [60]:
# Q7 Check for duplicate values
df_dups_ords = df_ords[df_ords.duplicated()]
df_dups_ords.size

0

No duplicates found

In [61]:
# Export data
# Export df_ords to prepared data
path=r'C:\Users\ponni\Achievement 4'
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_consistency checked.csv'))
# Exporting df_dep_t_new to prepared data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_consistency checked.csv'))